File: app/src/main/java/com/example/githubproject/ui/main/viewmodel/MainViewModel.java
package com.example.githubproject.ui.main.viewmodel;

import android.util.Log;

import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.githubproject.data.model.CarsModel;
import com.example.githubproject.data.repository.MainRepository;
import com.example.githubproject.utils.NetworkHelper;
import com.example.githubproject.utils.Resource;

import io.reactivex.Scheduler;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;

public class MainViewModel extends ViewModel {

    private MutableLiveData<Resource<CarsModel>> _users;
    private NetworkHelper networkHelper;
    private MainRepository mainRepository;

    @ViewModelInject
    public MainViewModel(NetworkHelper networkHelper, MainRepository mainRepository){
        this.networkHelper = networkHelper;
        this.mainRepository = mainRepository;
    }

    public LiveData<Resource<CarsModel>> getUsers(){
        if (_users == null){
            _users = new MutableLiveData<Resource<CarsModel>>();
            fetchUsers(1);
        }
        return _users;
    }

    public void fetchUsers(int page){
        _users.postValue(Resource.Companion.loading(null));
        Log.d("zzzz", "--- page: " + page);
        if (networkHelper.isNetworkConnected()) {
            mainRepository.getUsers(page).subscribeOn(Schedulers.io())
                    .subscribe(carsModelResponse -> {
                        if (carsModelResponse != null){
                            if (carsModelResponse.isSuccessful()) {
                                _users.postValue(Resource.Companion.success(carsModelResponse.body()));
                            } else _users.postValue(Resource.Companion.error(carsModelResponse.errorBody().toString(), null));
                        }
                    });
        } else _users.postValue(Resource.Companion.error("No internet connection", null));

    }

}

File: app/src/main/java/com/example/githubproject/ui/main/viewmodel/MainViewModel.kt
package com.example.githubproject.ui.main.viewmodel

import android.util.Log
import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.*
import com.example.githubproject.data.model.*
import com.example.githubproject.data.repository.MainRepository
import com.example.githubproject.utils.NetworkHelper
import com.example.githubproject.utils.Resource
import kotlinx.coroutines.launch

class MainViewModel @ViewModelInject constructor(
    private val networkHelper: NetworkHelper,
    private val mainRepository: MainRepository): ViewModel(){

    private val _users = MutableLiveData<Resource<WeatherModel>>()
    val users: LiveData<Resource<WeatherModel>> get() = _users

    private val _regions = MutableLiveData<Resource<List<Region>>>()
    val regions: LiveData<Resource<List<Region>>> get() = _regions

    init {
        fetchUsers("London")
    }

    fun fetchUsers(name: String){
        viewModelScope.launch {
            _users.postValue(Resource.loading(null))
            if (networkHelper.isNetworkConnected()) {
                mainRepository.getUsers("", name, 0, "", "").let {
                    if (it.isSuccessful) {
                        _users.postValue(Resource.success(it.body()))
                    } else _users.postValue(Resource.error(it.errorBody().toString(), null))
                }
            } else _users.postValue(Resource.error("No internet connection", null))
        }
    }

    fun fetchRegions(name: String){
        viewModelScope.launch {
            _regions.postValue(Resource.loading(null))
            if(networkHelper.isNetworkConnected()){
                mainRepository.getRegion("c4adb88e79e646ef95f213432221708", name).let {
                    if (it.isSuccessful){
                        _regions.postValue(Resource.success(it.body()))
                    }else {_regions.postValue(Resource.error(it.errorBody().toString(), null))}
                }

            }else _users.postValue(Resource.error("No internet connection", null))
        }
    }
}